src/live_executor.cc

```diff
@@ -0,0 +1,616 @@
+			if(unit_time < 1.0) { unit_time *= 1000.0, unit = "ns"; }
+			CELERITY_TRACE("[executor] retired I{} after {:.2f} {}", async.instr->get_id(), unit_time, unit);
+		} else {
+			CELERITY_TRACE("[executor] retired I{}", async.instr->get_id());
+		}
+	}
+
+	if(utils::isa<alloc_instruction>(async.instr)) {
+		const auto ainstr = utils::as<alloc_instruction>(async.instr);
+		const auto ptr = async.event.get_result();
+		assert(ptr != nullptr && "backend allocation returned nullptr");
+		const auto aid = ainstr->get_allocation_id();
+		CELERITY_TRACE("[executor] {} allocated as {}", aid, ptr);
+		assert(allocations.count(aid) == 0);
+		allocations.emplace(aid, ptr);
+	}
+
+	engine.complete_assigned(async.instr);
+}
+
+template <typename Instr>
+auto executor_impl::dispatch(const Instr& instr, const out_of_order_engine::assignment& assignment)
+    // SFINAE: there is a (synchronous) `issue` overload above for the concrete Instr type
+    -> decltype(issue(instr)) //
+{
+	assert(assignment.target == out_of_order_engine::target::immediate);
+	assert(!assignment.lane.has_value());
+	issue(instr); // completes immediately
+	engine.complete_assigned(&instr);
+}
+
+template <typename Instr>
+auto executor_impl::dispatch(const Instr& instr, const out_of_order_engine::assignment& assignment)
+    // SFINAE: there is an `issue_async` overload above for the concrete Instr type
+    -> decltype(issue_async(instr, assignment, std::declval<async_instruction_state&>())) //
+{
+	auto& async = in_flight_async_instructions.emplace_back();
+	async.instr = assignment.instruction;
+	issue_async(instr, assignment, async); // stores event in `async` and completes asynchronously
+}
+
+void executor_impl::try_issue_one_instruction() {
+	auto assignment = engine.assign_one();
+	if(!assignment.has_value()) return;
+
+	matchbox::match(*assignment->instruction, [&](const auto& instr) { dispatch(instr, *assignment); });
+	made_progress = true;
+}
+
+void executor_impl::check_progress() {
+	if(!policy.progress_warning_timeout.has_value()) return;
+
+	if(made_progress) {
+		last_progress_timestamp = std::chrono::steady_clock::now();
+		progress_warning_emitted = false;
+		made_progress = false;
+	} else if(last_progress_timestamp.has_value()) {
+		// being stuck either means a deadlock in the user application, or a bug in Celerity.
+		const auto elapsed_since_last_progress = std::chrono::steady_clock::now() - *last_progress_timestamp;
+		if(elapsed_since_last_progress > *policy.progress_warning_timeout && !progress_warning_emitted) {
+			std::string instr_list;
+			for(auto& in_flight : in_flight_async_instructions) {
+				if(!instr_list.empty()) instr_list += ", ";
+				fmt::format_to(std::back_inserter(instr_list), "I{}", in_flight.instr->get_id());
+			}
+			CELERITY_WARN("[executor] no progress for {:.3f} seconds, might be stuck. Active instructions: {}",
+			    std::chrono::duration_cast<std::chrono::duration<double>>(elapsed_since_last_progress).count(),
```

Comment:
I always find this thing horrible to read compared to
```C++
using std::chrono_literals;
// ...
elapsed_since_last_progress / 1s
```
I think there are only 2 occurrences in this TU, so maybe it's not worth it to switch, but I do think it's a lot more readable.

Reply:
`/ 1s` gives me `long` and `/ 1.0s` gives me `long double` (??) I don't like it.

I have decided to overengineer this instead, cf `print_utils.h`.

Reply:
I did mean `1.0s`, I'm not sure why it matters that it's a `long double` in this case.